pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/ComplexTypeTransformer.java

```diff
@@ -0,0 +1,184 @@
+  }
+
+  public static boolean isComplexTypeHandlingEnabled(TableConfig tableConfig) {
+    if (tableConfig.getIngestionConfig() == null
+        || tableConfig.getIngestionConfig().getComplexTypeHandlingConfig() == null
+        || tableConfig.getIngestionConfig().getComplexTypeHandlingConfig().getMode() == null) {
+      return false;
+    }
+    return tableConfig.getIngestionConfig().getComplexTypeHandlingConfig().getMode()
+        != ComplexTypeHandlingConfig.Mode.NONE;
+  }
+
+  @Nullable
+  @Override
+  public GenericRow transform(GenericRow record) {
+    flattenMap(record, new HashSet<>(record.getFieldToValueMap().keySet()));
+    for (String collection : _collectionsToUnnest) {
+      unnestCollection(record, collection);
+    }
+    return record;
+  }
+
+  private GenericRow unnestCollection(GenericRow record, String column) {
+    if (record.getValue(GenericRow.MULTIPLE_RECORDS_KEY) == null) {
+      List<GenericRow> list = new ArrayList<>();
+      unnestCollection(record, column, list);
+      record.putValue(GenericRow.MULTIPLE_RECORDS_KEY, list);
+    } else {
+      Collection<GenericRow> records = (Collection) record.getValue(GenericRow.MULTIPLE_RECORDS_KEY);
+      List<GenericRow> list = new ArrayList<>();
+      for (GenericRow innerRecord : records) {
+        unnestCollection(innerRecord, column, list);
+      }
+      record.putValue(GenericRow.MULTIPLE_RECORDS_KEY, list);
+    }
+    return record;
+  }
+
+  private void unnestCollection(GenericRow record, String column, List<GenericRow> list) {
+    Object value = record.removeValue(column);
+    if (value == null) {
+      // use the record itself
+      list.add(record);
+      return;
+    } else if (value instanceof Collection) {
+      if (((Collection) value).isEmpty()) {
+        // use the record itself
+        list.add(record);
+      } else {
+        for (Object obj : (Collection) value) {
+          GenericRow copy = flattenCollectionItem(record, obj, column);
+          list.add(copy);
+        }
+      }
+    } else if (value.getClass().isArray()) {
+      if (((Object[]) value).length == 0) {
+        // use the record itself
+        list.add(record);
+      } else {
+        for (Object obj : (Object[]) value) {
+          GenericRow copy = flattenCollectionItem(record, obj, column);
+          list.add(copy);
+        }
+      }
+    }
+  }
+
+  private GenericRow flattenCollectionItem(GenericRow record, Object obj, String column) {
+    GenericRow copy = record.copy();
+    if (obj instanceof Map) {
+      Map<String, Object> map = (Map<String, Object>) obj;
+      for (Map.Entry<String, Object> entry : new HashSet<>(map.entrySet())) {
+        String flattenName = concat(column, entry.getKey());
+        copy.putValue(flattenName, entry.getValue());
+      }
+    } else {
+      copy.putValue(column, obj);
+    }
+    return copy;
+  }
+
+  @VisibleForTesting
+  protected GenericRow flattenMap(GenericRow record, Collection<String> columns) {
```

Comment:
Why is this not private ?
This should be called internally from `transform()` or `unnestCollection()` after determining we have to flatten the map.

Reply:
used for testing to make the test more granular.

Reply:
Do we need to return a `GenericRow` here? It is always in-place modification